app/scripts/lib/metaRPCClient.js

```diff
@@ -0,0 +1,77 @@
+import { EventEmitter } from 'events';
+import { EthereumRpcError } from 'eth-rpc-errors';
+import createRandomId from '../../../shared/modules/random-id';
+
+class MetaRPCClient {
+  constructor(connectionStream) {
+    this.connectionStream = connectionStream;
+    this.notificationChannel = new EventEmitter();
+    this.requests = new Map();
+    this.connectionStream.on('data', this.handleResponse.bind(this));
+    this.connectionStream.on('end', this.close.bind(this));
+    // eslint-disable-next-line no-constructor-return
+    return new Proxy(this, {
```

Comment:
It would be a bit more conventional to use a factory method and add the wrapping proxy there, rather than returning a proxy from a constructor. I've never even seen this done before, and I can't find any information about whether this is spec-compliant or whether it works the same on all browsers.